Use GitSCMFileSystem caches as reference repository for GitSCM clone

atikhono opened the pull request:
[JENKINS-43894](https://issues.jenkins-ci.org/browse/JENKINS-43894) - Environment variables not resolved in Pipeline SCM -> Advanced clone behaviours -> Path of the reference repo

Implement automatic caching of git repositories as discussed in the comments. Extend GitSCM to automatically use GitSCMFileSystem cache as a reference repository. Update such caches before any checkout. Move the cache implementation out of AbstractGitSCMSource to reuse in GitSCM. Implement cache retrieval for remote slaves. Use the new cache implementation in GitSCMFileSystem.

@MarkEWaite @jglick could you please have a look?

jglick commented:
> Extend GitSCM to automatically use GitSCMFileSystem cache as a reference repository.

This is probably dangerous—we would want it to be opt-in.

MarkEWaite commented:
Jesse, can you explain further why you consider it dangerous to automatically use the cache as a reference repository?  You said:

> > Extend GitSCM to automatically use GitSCMFileSystem cache as a reference repository.

> This is probably dangerous—we would want it to be opt-in.

As far as I understand, command line git treats a reference repository as a cache to search for content before requesting it over the wire.  I believe the content it is searching is SHA1 hashed content so it has no worse risk of information leaks than any other content in the repository.

jglick commented on src/main/java/hudson/plugins/git/Cache.java:
```diff
@@ -0,0 +1,92 @@
+            LOGGER.severe("Jenkins instance is null in Cache.getCacheDir");
+            return null;
+        }
+        File cacheDir = new File(new File(jenkins.getRootDir(), "caches"), getCacheEntry(remoteURL));
```
This seems to be exact same logic as above, but with an implicit `node` of `Jenkins.getInstance()`.

atikhono replied:
Yes, but one is File (for master) and another FilePath (for other nodes). I didn't want to mess with GitSCMFileSystem code too much, so I left the getCacheDir implementation as is for master.

jglick commented on src/main/java/hudson/plugins/git/GitSCM.java:
```diff
@@ -1106,7 +1143,7 @@ public void checkout(Run<?, ?> build, Launcher launcher, FilePath workspace, Tas
         }
 
-        retrieveChanges(build, git, listener);
+        retrieveChanges(build, git, listener, GitUtils.workspaceToNode(workspace));
```
Can `workspaceToNode` return null?

atikhono replied:
Well, it uses Jenkins.getActiveInstance which is Nonnull... But the 2 tests my patch is failing seem to have an NPE caused by a null result of Jenkins.getInstance (which is Nonnull in the API reference too).

jglick replied:
If `Jenkins.getInstance()` returns null it means that you are making the call from inside an agent JVM, signalling a fundamental design mistake somewhere.

atikhono replied:
That's how it was implemented in AbstractGitSCMSource.getCacheDir. There's a check for null Jenkins.getInstance. Now I'm using node to determine the cache directory to lock as well, so I need to have this check in my code. I've added it.

jglick commented:
> Extend GitSCM to automatically use GitSCMFileSystem cache as a reference repository.

This is actually a bit misleading, because they will not be using the _same_ cache. `GitSCMFileSystem` operates on the master. On a properly configured system, `GitSCM` used in the normal way during a build operates on an agent.

What is being shared is the cache _code_, not the caches themselves as the Mercurial plugin does. (More precisely, the Mercurial plugin in the “caching” mode maintains two levels on caches—on the master and on agents; it uses bundle transfers from master to agent where needed, and then uses a local clone/update on the agent from the agent’s cache. Optionally it also uses the “sharing” extension, which I think is the closest equivalent to Git’s reference repositories.) Here a cache will be shared only between different builds (or checkouts, more generally) running on the same agent.

> why you consider it dangerous

Just as a general principle—this is proposing a significant behavioral change that would affect most users of the Git plugin, hence most Jenkins users. There is a high risk of unforeseeable regressions.

There are also some potential downsides that may affect some users more than others. For example, a job using a gigantic repository and performing aggressive workspace cleanup on permanent agents would begin to see disk usage not drop much after such cleanups, because the caches would remain. And on a “one-shot executor” creation of the cache from scratch on each build is simply added complexity and a slight performance hit.

Yes these considerations potentially apply also to multibranch projects using Git (and not GitHub), or generally `GitSCMFileSystem`, but those use cases are opt-in. With this PR, someone simply applying a routine plugin upgrade would see the change.

MarkEWaite commented:
Thanks for your insights.

> > why you consider it dangerous
> Just as a general principle—this is proposing a significant behavioral change that would affect most users of the Git plugin, hence most Jenkins users. There is a high risk of unforeseeable regressions.

That's good advice.  When I review the pull request, I'll be looking carefully for ways that we can introduce the capability without breaking compatibility for the 100 000+ installations using the plugin.

atikhono commented:
Added an option to the global plugin config (useCaches).
Added a check for null instance in Cache.lock and Cache.unlock to let correctGitToolIsUsed and correctGitToolIsUsed2 tests pass.

atikhono commented on src/main/resources/hudson/plugins/git/GitSCM/help-useCaches.html:
```diff
@@ -0,0 +1,11 @@
+<div>
+    <p>When checked, Hudson will maintain a clone of the Git repository
+    in use as a reference for each clone of that repository. The clone
+    will be kept on each node and updated before any other clone of the
```
Wow. Sorry for that. I will fix the description when I have something meaningful to fix. Just not to waste the CI for this.

sgtcoolguy commented:
Not to spam the comments, but for reference I had filed a ticket related directly to this on Jenkins JIRA: https://issues.jenkins-ci.org/browse/JENKINS-44729

I'd love to see some improvement here, since for some larger popular repos we're getting 75+ clones of the same base repo on every build agent, which is a monumental disk/network waste. Just having a per-agent reference repo/cache would make a huge difference.

atikhono commented:
@jglick @MarkEWaite up
Rebased on the latest master

atikhono commented:
I've rebase the patch
@jenkinsci/code-reviewers please have a look

atikhono commented:
Wow. Hit the Close button by mistake.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

MarkEWaite commented:
I applied changes for a few of the items detected by @oleg-nenashev (thanks very much, Oleg!), added a test of the Cache class, and pushed it to a [development branch](https://github.com/MarkEWaite/git-plugin/commits/master-PR502-use-cache) that I'll use for interactive evaluation.  I am unlikely to work on the interactive validation until next week.

I can submit those proposed changes as a pull request to your branch, or you can evaluate them and decide if you prefer to make those changes some other way.

atikhono commented on src/main/java/hudson/plugins/git/Cache.java:
```diff
@@ -0,0 +1,102 @@
+        ConcurrentMap<String, Lock> cacheNodeLocks;
+        Lock cacheLock;
+        String nodeEntry = getNodeEntry(node);
+        while (null == (cacheNodeLocks = cacheLocks.get(nodeEntry))) {
```
They are created only if an item is missing in a corresponding hashMap. Not sure I got the question, through.

atikhono commented on src/main/java/hudson/plugins/git/Cache.java:
```diff
@@ -0,0 +1,102 @@
+        if (jenkins != null) {
+            lock(jenkins, remoteURL);
+        } else {
+            LOGGER.severe("Jenkins instance is null in Cache.lock");
```
Yep, will do

atikhono replied:
I will have to put all Cache.lock/Cache.unlock calls under try-cacthes and write catch branches for a thrown exception. From Mark's patch ([cdfdf17](https://github.com/MarkEWaite/git-plugin/commit/bb47cdfdf174b9c6e8e057ca17796a5677d42ef9)) I can see the null check isn't necessary after upgrade to 2.60 core. Does it mean that this new exception handling code should be removed then as well?

atikhono replied:
@oleg-nenashev this method has been moved from AbstractGitSCMSource untouched: [link](https://github.com/jenkinsci/git-plugin/blob/master/src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java#L670). If I throw an exception from there, I'll have two failing tests (unexpected exception):
- AbstractGitSCMSourceRetrieveHeadsTest.correctGitToolIsUsed
- AbstractGitSCMSourceRetrieveHeadsTest.correctGitToolIsUsed2
I cannot suppress cache use for these tests because it uses master cache in GitSCMFileSystem. I'm not sure if it's a test issue, and don't see the right way to fix it. Or should  I fix it in scope of this change at all.

atikhono commented:
@oleg-nenashev Thanks very much for the review!
@MarkEWaite I've cherry-picked your commits to my local branch. When I'm done fixing what's left, I can either put my updates on top of that branch and update the pull request (so that we can see just the updates). Or I can squash all commits into one. Which way is the better for review?

MarkEWaite commented:
@atikhono I'd prefer the commits be separate rather than squashed.

My interactive testing a day or two ago encountered several surprising failures in my collection of agents on various platforms.  I don't have specific enough details to report anything that you are likely to duplicate in your environment.

Some of the oddities may be testable for you from my [lts-with-plugins docker image](https://github.com/MarkEWaite/docker-lfs/tree/lts-with-plugins), if you add one or more agents.

An example of the failure case I saw was in a multi-branch pipeline job from a folder that is loading a pipeline library.  The library failed to clone, with a message

    error: https://github.com/MarkEWaite/jenkins-pipeline-utils did not send all necessary objects

The greater log looks like this:
```
Obtained Jenkinsfile from dcceb8e61f760fea60bc2b02a9e9b05b311bd361
Loading library globalPipelineLibraryMarkEWaite@master
Cloning the remote Git repository
Creating cache in /var/jenkins_home/caches/git-27bcd32d21ed6c836389204979423813
Using shallow clone
Avoid fetching tags
Honoring refspec on initial clone
Cloning repository https://github.com/MarkEWaite/jenkins-pipeline-utils

[content removed that didn't seem relevant]

Cloning repository https://github.com/MarkEWaite/jenkins-pipeline-utils
 > git init /var/jenkins_home/jobs/Bugs-Pipeline-Checks/jobs/jenkins-bugs-multibranch-pipeline/branches/JENKINS-34309/workspace@libs/globalPipelineLibraryMarkEWaite # timeout=10
Fetching upstream changes from https://github.com/MarkEWaite/jenkins-pipeline-utils
 > git --version # timeout=10
using GIT_ASKPASS to set credentials intentionally invalid user and password for credentials NPE in multi-branch jobs
 > git fetch --no-tags --progress https://github.com/MarkEWaite/jenkins-pipeline-utils +refs/heads/${library.globalPipelineLibraryMarkEWaite.version}:refs/remotes/origin/${library.globalPipelineLibraryMarkEWaite.version} --depth=1
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Command "git fetch --no-tags --progress https://github.com/MarkEWaite/jenkins-pipeline-utils +refs/heads/${library.globalPipelineLibraryMarkEWaite.version}:refs/remotes/origin/${library.globalPipelineLibraryMarkEWaite.version} --depth=1" returned status code 1:
stdout: 
stderr: remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0        
fatal: bad object 0c30065c158df07e55eeda283a7db3ff19bbfe01
error: https://github.com/MarkEWaite/jenkins-pipeline-utils did not send all necessary objects


	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1924)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1643)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:71)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:352)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:559)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1110)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1163)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:113)
	at org.jenkinsci.plugins.workflow.libs.SCMSourceRetriever.doRetrieve(SCMSourceRetriever.java:108)
	at org.jenkinsci.plugins.workflow.libs.SCMRetriever.retrieve(SCMRetriever.java:64)
	at org.jenkinsci.plugins.workflow.libs.LibraryAdder.retrieve(LibraryAdder.java:150)
	at org.jenkinsci.plugins.workflow.libs.LibraryAdder.add(LibraryAdder.java:131)
	at org.jenkinsci.plugins.workflow.libs.LibraryDecorator$1.call(LibraryDecorator.java:114)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1065)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:603)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:581)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:558)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)
	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)
	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.doParse(CpsGroovyShell.java:129)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:123)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:516)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:479)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:268)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:405)
ERROR: Error cloning remote repo 'origin'
```

sgtcoolguy commented:
@MarkEWaite It seems like maybe it's not using the string interpolation to selective branch for the clone/fetch?

Unless you actually have a branch named `${library.globalPipelineLibraryMarkEWaite.version}`?

```
https://github.com/MarkEWaite/jenkins-pipeline-utils +refs/heads/${library.globalPipelineLibraryMarkEWaite.version}:refs/remotes/origin/${library.globalPipelineLibraryMarkEWaite.version}
```

atikhono commented:
@MarkEWaite The issue was in concurrent access to cache objects, because a proper cache didn't get locked. Could you please run your tests once again?

MarkEWaite commented:
@atikhono I'll build it and review it.

atikhono commented:
rebased
up @reviewbybees @MarkEWaite any updates?

MarkEWaite commented:
I've been testing it in my environment for about a week.  There are several failures that I've not had time to describe in more detail.

One failure seems to be specific to older git versions (as delivered on CentOS 7, and CentOS 6, and possibly on Debian 7).  Those git versions often report that they cannot clone from a shallow repository.  I interpret that to mean that the cached copy may be (correctly) honoring a request for a shallow clone, but if the git version is older than git 1.8, it probably should ignore the request for shallow clone.

Another failure that I've seen in logs but which has not appeared as a job failure seems to be related to a damaged copy of a repository being in the cache.  The checkout seems to attempt to use the damaged cache, fails, and then (correctly) ignores the error and continues with the checkout.  I'm not sure if anyone else will see that problem, since I don't see the problem if I discard the workspaces and the caches and restart.  I assume that failure is due to the previous version with the locking that was not tied to a specific agent.

I need to investigate other log files more deeply to identify other areas which the testing has uncovered.  The code with these changes definitely fails builds in my test infrastructure more frequently than the current code.  I believe it has also reduced my disc use and improved performance.  I don't have measurements yet to confirm that, but hope to gather them.

traskrogers commented:
Any updates regarding this PR?  This would cleanly and transparently improve our disk-space issues from all the duplicated branches since moving a bunch of projects to multi-branch pipelines.
